### PR TITLE
Cirrus: Task consolidation

### PR DESCRIPTION
**Pull request by cevich**
The set of Cirrus-CI tasks seems to be ever-expanding, in addition to a tangled dependency-tree.  This is slowing down the overall CI process slightly (due to task queuing) and adds to the mental load required to service the YAML.  Fix both by consolidating several semi-related tasks into a slightly larger build-task.

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by cevich** [comment]
Woopsie this isn't ready yet....still need to do something with the validate and consistency checks.

**Comment by cevich** [comment]
This is ready for review.

**Comment by rhatdan** [comment]
@lsm5 @edsantiago PTAL

**Comment by edsantiago** [comment]
New task map: 
cirrus-map-pr15782
This is way too complicated for me to review in the next half-hour, sorry. It'll have to be after f2f.

**Comment by cevich** [comment]
> This is way too complicated for me to review in the next half-hour, sorry. It'll have to be after f2f.

It's not a problem.  It's an increase in YAML & script complexity in in favor of questionably faster CI.  I wanted to roll the `Validate` task (and a few others) up into `Build`, but that was even more complex :disappointed:   Perhaps this can be done in a future PR.

File: contrib/cirrus/CIModes.md
Comment (lsm5) [inline contrib/cirrus/CIModes.md:113]: @cevich just doublechecking that you intended to remove the build task from the branch task list.

**Review by lsm5 — APPROVED** [review]
LGTM. I :heart:  less yaml in favour of more scripts :D

File: contrib/cirrus/shellcheck.sh
Comment (edsantiago) [inline contrib/cirrus/shellcheck.sh:8]: I'm frantically identifying all the appropriate diffs to run. As part of that, I noticed that this `cd` was removed from the final `prebuild.sh`. Was that intentional? Is there a working-directory guarantee? I couldn't find one in `.cirrus.yml`.

File: contrib/cirrus/prebuild.sh
Comment (edsantiago) [inline contrib/cirrus/prebuild.sh:73]: This (`-e`) is the next diff I see. Just curious, what does that gain us?

**Comment by edsantiago** [comment]
(not included in the diff, because it's too far out of scope): could you please remove `ext_svc` and `automation` from this line?
https://github.com/containers/podman/blob/828fae12971c5a7b9807c8c4f8e029fe5d0ddc2f/.cirrus.yml#L55

File: contrib/cirrus/prebuild.sh
Comment (edsantiago) [inline contrib/cirrus/prebuild.sh:29]: Looking at the logs, I just see a jumble of output messages with no indication of what is what. I don't think we want `set -x` here, but what would you think of adding: `echo "Running cirrus_yaml_test"`, `echo "running shellcheck"`, so when one of those steps fails it's easy to see what it actually was?

File: contrib/cirrus/postbuild.sh
Comment (edsantiago) [inline contrib/cirrus/postbuild.sh:22]: I'm not so sure this `skip` is a good idea. Did you look at the logs? Did you see the humongous diff between f36 [intel](https://cirrus-ci.com/task/5841637799624704?logs=postbuild#L13) and [aarch64](https://cirrus-ci.com/task/6404587753046016?logs=postbuild#L13)? Given the sometimes significant differences in `go` between distros, I think we would be wise to run this check everywhere.

File: contrib/cirrus/postbuild.sh
Comment (edsantiago) [inline contrib/cirrus/postbuild.sh:25]: Another place where `echo running make vendor / generate-bindings / completions` progress messages would be appreciated by a future hapless investigator.

**Comment by edsantiago** [comment]
Whew. Okay, I think I'm done. LGTM with a few suggestions, make of them what you will. Thanks for the cleanup!

File: contrib/cirrus/CIModes.md
Comment (cevich) [reply contrib/cirrus/CIModes.md:113]: Oh woops, no I didn't.  Thanks for catching that.

File: contrib/cirrus/shellcheck.sh
Comment (cevich) [reply contrib/cirrus/shellcheck.sh:8]: I'm not sure if it's documented anywhere, but it has always behaved this way.  In any case, it probably wouldn't hurt to put this into the new scripts just to be safe and obvious to maintainers.

File: contrib/cirrus/prebuild.sh
Comment (cevich) [reply contrib/cirrus/prebuild.sh:73]: It catches the unlikely/off-chance that quay returns invalid, "false", or null JSON.  Sadly it won't catch "empty" output, so maybe that's something I should add.

File: contrib/cirrus/prebuild.sh
Comment (cevich) [reply contrib/cirrus/prebuild.sh:29]: Oh sure, didn't realize/remember this was `set -x`.

File: contrib/cirrus/postbuild.sh
Comment (cevich) [reply contrib/cirrus/postbuild.sh:22]: okay, fair enough.

File: contrib/cirrus/postbuild.sh
Comment (cevich) [reply contrib/cirrus/postbuild.sh:25]: Yep, I've got a library function `showrun()` that behaves like single-line `set -x`.  That should do the trick.

**Comment by cevich** [comment]
> 
Whew. Okay, I think I'm done. LGTM with a few suggestions, make of them what you will. Thanks for the cleanup!

Thanks for your scrutiny, and @lsm5 too!

File: contrib/cirrus/prebuild.sh
Comment (edsantiago) [reply contrib/cirrus/prebuild.sh:73]: Correction: it catches "false" or "null" only; invalid JSON is already caught by plain `jq`:
```console
$ echo hi | jq .
parse error: Invalid numeric literal at line 2, column 0
```
I guess I'm ok with it, but am having trouble imagining what possible situation it will detect.

**Comment by edsantiago** [comment]
Sorry Dan, this is not ready; it's pending fixes.

**Comment by rhatdan** [comment]
Sorry about that,

**Comment by edsantiago** [comment]
@cevich quick reminder, this is in your ballpark

**Comment by cevich** [comment]
> this is in your ballpark

Yep, I haven't forgotten.  Just letting it marinade a bit :grin:

File: contrib/cirrus/prebuild.sh
Comment (cevich) [reply contrib/cirrus/prebuild.sh:73]: It's just belt-and-suspenders defensive-coding.  I'd rather it catch an extremely rare and fantastically unexpected result than let it slide as "normal".

Comment (cevich) [reply contrib/cirrus/prebuild.sh:73]: I'll add a comment.

**Comment by cevich** [comment]
Force-push: Applied all suggestions, rebased on main.

**Comment by cevich** [comment]
force-push: mid-air collision, rebased and resolved conflicts.

**Comment by cevich** [comment]
force-push: Squashed fixup commit.

File: contrib/cirrus/postbuild.sh
Comment (cevich) [reply contrib/cirrus/postbuild.sh]: Ugh...[I was just reminded why I added this](https://cirrus-ci.com/task/5070067325992960).  We need to exclude Ubuntu.

**Comment by cevich** [comment]
Force-push: Added shellcheck skip for Ubuntu.

**Comment by cevich** [comment]
Force-push: Shellcheck fixups.

**Comment by cevich** [comment]
Force-push: Rebased & resolved conflicts.

**Comment by edsantiago** [comment]
LGTM but the `SC2154` bug is triggering. (As in, bug in the linter, which is not set up to understand our shared code base)

**Comment by cevich** [comment]
Oof, maybe I should just disable it.  We do have and use the `req_env_vars` function anyway, so the SC2154 thing is less useful in this repo.

**Comment by edsantiago** [comment]
> We do have and use the req_env_vars function anyway

That doesn't help against typos:
```bash
req_env_vars SCRIPT_BASE
... $SCRITP_BASE
```
...but then, neither does shellcheck.

/lgtm
/hold

I reviewed using `range-diff`, not the thorough very-long review I did on my first pass; I'm puzzled by putting `check_go_changes` in `postbuild` (it seems to make more sense in `prebuild`, at least to me) but that's not important enough to block on.

**Comment by cevich** [comment]
> I'm puzzled by putting check_go_changes in postbuild

FWIW, my decision was based on the (possibly tenuous) argument: It *Compiles*, is slightly more important than, it *looks good*.  Assuming current and future `check_go_changes` checks are all functionally equivalent.

**Pull request merged**
